### DISCUSSION TITLE
add cheats

### DISCUSSION THREAD

**Discussion by JurajHanakovic** [ihhub/fheroes2#5265] [General]
This game/sourceport/engine-recreation really needs to add cheats.

Some scenarios are extra hard for casual and non-casual player alike. Especially when playing in non-optimal ways.
And I don't want to replay 2h+ scenario if I don't play optimaly all the time.
You know, games are supposed to be fun, not a chore to play.

I'm quite certain that many other players will appreciate cheat options to be added.

**Comment by zenseii** [comment]
Hi @JurajHanakovic ! Fheroes2 is a complete engine recreation of Homm2, so the fact that there are no cheats is not because they aren't "enabled", but because they simply haven't been implemented in the code. The team is prioritizing improving the game rather than make native cheats.

The AI is also just in its infancy and will only continue to get improvements and get better. My suggestion is to look around for tips and tricks about the game and maybe watch some videos of people playing it. You're also welcome to join the Discord channel to ask for help and advice there. You can also ask here of course, but less people come here to discuss.

**Reply by JurajHanakovic** [reply]
Thx for reply, but...

The issue is not that I don't know HOW to play the game (I played most of HOMAM games throughout my life and got quite good at them). I just like to play it in DIFFERENT non-optimal ways to have more fun.

Implementing cheats isn't that hard. Honestly, I could probably do it in a few hours if I'm familiar with this project's code and I'm just a sub-par programmer.

I'm currently using workaround with CheatEngine but it's really not ideal solution.
Also, I'm quite certain that many other players will appreciate cheat options to be added.

**Reply by zenseii** [reply]
> Thx for reply, but...
> 
> The issue is not that I don't know HOW to play the game (I played most of HOMAM games throughout my life and got quite good at them). I just like to play it in DIFFERENT non-optimal ways to have more fun.
> 
> Implementing cheats isn't that hard. Honestly, I could probably do it in a few hours if I'm familiar with this project's code and I'm just a sub-par programmer.
> 
> I'm currently using workaround with CheatEngine but it's really not ideal solution. Also, I'm quite certain that many other players will appreciate cheat options to be added.

Not much I can say other than that this is indeed open-source and I'm sure the core team would appreciate help. There aren't many people working regularly on this project so efforts have to be focused on more fundamental things to the experience.

**Reply by ihhub** [reply]
Hi @JurajHanakovic , as pointed by other people let's fix the root of the problem rather than fixing consequences. If you want to enjoy the game on a full scale you must understand the rules of it. Playing chess without playing by rules is the same. Alternatively, you're highly encouraged to play on Easy mode

Implementing cheats is not straightforward: how would they work on PlayStation Vita or future Android release? A feature must work on every platform we support.

**Reply by JurajHanakovic** [reply]
Hi, ihhub.
Read my reply to zenseii. Your point of me not knowing how to play this game is incorrect and irrelevant.

But ok, if you/others don't want to add easily accessible cheats (without debug build etc), it's your decision.

**Reply by ihhub** [reply]
Hi @JurajHanakovic , the point here is that we have something for testing highlighted by @oleg-derevenetz . Implementing cheats is not an easy task and it has lowest priority right now.

**Comment by oleg-derevenetz** [comment]
Hi @JurajHanakovic there are already cheats, they are just called "debug mode". You can build `fheroes2` in debug mode (for example, using the `Debug` configuration in CMake) and then you will be able to have fun with the `debug` setting in the `fheroes2.cfg`. For example, you will be able to view the entire map or get a mighty `Debug Hero` with all spells and black dragons. In any case, you are able to alter any logic in the game as you like, because this game is open source. I make up such parameters of heroes, castles and even individual units that I need during debugging on a regular basis. All those cheats are nothing compared to what you can do by making changes to the sources.

**Reply by JurajHanakovic** [reply]
Hi, Oleg.
Debug build is indeed an option. Thanks for pointing that out. It can be what I'm looking for... 
Still, it's not very convenient for average user/player.
Anyway, thanks for reply. I'll look into this solution.

**Comment by ihhub** [comment]
This discussion is similar to #2282. To do not have multiple topics we are closing this discussion.
